keps/sig-api-machinery/4020-unknown-version-interoperability-proxy/README.md

```diff
@@ -0,0 +1,989 @@
+   5. If the local apiserver ID is found in the list of serviceable-by server IDs from the internal map
+
+3. If the local apiserver ID is not found in the list of serviceable-by server IDs, a random apiserver ID will be selected from the retrieved list and the request will be proxied to this apiserver
+
+4. If there is no apiserver ID retrieved for the requested GVR, we will serve 404 with error `GVR <group_version_resource> is not served by anything in this cluster`
+
+5. If the proxy call fails for network issues or any reason, we serve 503 with error `Error while proxying request to destination apiserver`
+
+[handler chain]:https://github.com/kubernetes/kubernetes/blob/fc8f5a64106c30c50ee2bbcd1d35e6cd05f63b00/staging/src/k8s.io/apiserver/pkg/server/config.go#L639
+
+#### StorageVersion enhancement needed
+
+StorageVersion API currently tells us whether a particular StorageVersion can be read from etcd by the listed apiserver. We will enhance this API to also include apiserver ID of the server that can serve this StorageVersion.
+
+With the enhancement, the new [ServerStorageVersion](https://github.com/kubernetes/kubernetes/blob/release-1.27/pkg/apis/apiserverinternal/types.go#L62-L73) object will have this structure
+
+```
+type ServerStorageVersion struct {
+  // The ID of the reporting API server.
+  APIServerID string
+	
+  // The API server encodes the object to this version 
+  // when persisting it in the backend (e.g., etcd).
+  EncodingVersion string
+
+  // The API server can decode objects encoded in these versions.
+  // The encodingVersion must be included in the decodableVersions.
+  DecodableVersions []string
+
+  // Versions that can be served by the reporting API server
+  ServedVersions []string
+}
+```
+
+#### Identifying destination apiserver's network location
+
+We have listed the approaches to store/retrieve apiservers' ip and port information in the [Alternatives](#alternatives) section.
+
+We will use an egress dialer for network connections made to peer kube-apiservers. For this, will create a new type for the network context to be used for peer kube-apiserver connections ([xref](https://github.com/kubernetes/kubernetes/blob/release-1.27/staging/src/k8s.io/apiserver/pkg/apis/apiserver/types.go#L56-L71))
+
+#### Proxy transport between apiservers and authn
+
+For the mTLS between source and destination apiservers, we will do the following
+
+1. For server authentication by the client (source apiserver) : the client needs to validate the [server certs](https://github.com/kubernetes/kubernetes/blob/release-1.27/staging/src/k8s.io/apiserver/pkg/server/options/serving.go#L59) (presented by the destination apiserver), for which it will 
+   1. look at the CA bundle of the authority that signed those certs. We will introduce a new flag --peer-ca-file that **is required be passed (for this feature)** to the kube-apiserver that will be used to verify the presented server certs
```

Comment:
This is a better flag name:

```suggestion
   1. look at the CA bundle of the authority that signed those certs. We will introduce a new flag --requestheader-serving-ca-file that **is required be passed (for this feature)** to the kube-apiserver that will be used to verify the presented server certs
```

Reply:
I think we were planning to use a common "peer" prefix for the kube-apiserver <-> kube-apiserver comm. flags. As suggested by @jpbetz But can change it to "requestheader-serving-ca-file" if that makes more sense here.

Reply:
@Richabanker To help everyone on this comment understand to understand the full context. What is the full set of flags that will be involved in configuring `kube-apiserver <-> kube-apiserver` traffic?

Reply:
for authn:

--proxy-client-cert|key-file (presented by source apiserver to authn itself to destination apiserver)
--requestheader-client-ca-file (used by destination apiserver to authn source apiserver)

--tls-cert|(private-key)-file (presented by destination apiserver to authn itself to source apiserver)
--peer-ca-file (used by source apiserver to authn destination apiserver)

for locating the apiserver: 

--peer-bind-address

Reply:
> for authn:
>
>--proxy-client-cert|key-file (presented by source apiserver to authn itself to destination apiserver)
--requestheader-client-ca-file (used by destination apiserver to authn source apiserver)
>
>secure serving flag : --tls-cert|(private-key)-file (presented by destination apiserver to authn itself to source apiserver)
--peer-ca-file (used by source apiserver to authn destination apiserver)
>
>for locating the apiserver:
>
>--peer-bind-address

Probably worth an example in the doc.  You'll also need `--requestheader-allowed-names` I think.

Reply:
`--requestheader-serving-ca-file` still sounds better to me but I suspect any name we choose here will be difficult to explain in all contexts.

Reply:
I chatted with @enj about this. It sounds like it should work fine have the flags as described in this KEP.

I don't have a strong opinion on the name of the CA file flag.  It would be good to use consistent terms as much as possible across user visible surfaces (flags, HTTP header, API fields/label keys).  I'm okay with continuing naming discussions post-enhancement freeze (names selected in KEPs are not usually considered final).